lightning/src/routing/router.rs

```diff
@@ -1058,6 +1071,26 @@ impl<'a> CandidateRouteHop<'a> {
 			_ => None,
 		}
 	}
+	/// Returns the source node id of this hop.
+	pub fn source(&self) -> NodeId {
+		match self {
+			CandidateRouteHop::FirstHop { node_id, .. } => *node_id,
+			CandidateRouteHop::PublicHop { info, .. } => info.channel.node_one.into(),
```

Comment:
We can't assume this is `node_one`. For `ChannelInfo`, there is a canonical ordering of the nodes to determine which is `node_one` and which is `node_two`. But when routing, we want the `source` to be where the payment is coming from and the `target` to be where the payment is going to. Thus, why we have a `DirectedChannelInfo` struct. We may need to augment `DirectedChannelInfo` to keep track of this in some manner.